aten/src/ATen/test/undefined_tensor_test.cpp

```diff
@@ -0,0 +1,52 @@
+#include "ATen/ATen.h"
+#include <string>
+
+using namespace at;
+
+#define ASSERT_THROWS(fn, message)                                  \
+try {                                                               \
+  fn;                                                               \
+  assert(false);                                                    \
+} catch(std::runtime_error &e) {                                    \
+  assert(std::string(e.what()).find(message) != std::string::npos); \
+}
+
+
+int main() {
+  // mainly test ops on undefined tensors don't segfault and give a reasonable errror message.
+  Tensor und;
+  Tensor ft = CPU(kFloat).ones({1});
+
+  std::cout << und << std::endl;
+  assert(!und.defined());
+  assert(std::string("UndefinedTensor") == und.toString());
+
+  ASSERT_THROWS(und.strides(), "strides");
+  ASSERT_THROWS(und.dim(), "dim");
+  ASSERT_THROWS(und.assign_(Scalar(5)), "assign");
+  ASSERT_THROWS(und.unsafeGetTH(true), "unsafeGetTH");
+  ASSERT_THROWS(und.add(und), "add");
+  ASSERT_THROWS(und.add(ft), "add");
+  ASSERT_THROWS(ft.add(und), "add");
+  ASSERT_THROWS(und.add(5), "add");
+  ASSERT_THROWS(und.mm(und), "mm");
+
+  und.toType(und.type());
```

Comment:
Why do we allow these calls instead of throwing?

Reply:
It's arbitrary but doesn't seem unreasonable.  The test is just there to check it doesn't segfault and to alert if we change the behavior.

Reply:
What I was originally thinking: if we had the full undefined type lattice of {Dense/Sparse}{CPU/GPU}{Float/Double/Int}, you should be able to move along those as long as you don't go from Undefined to not or visa versa along one of those axes.  So this fits in with that model.  But let me know if you think the other way is better, I don't think it really matters.